test/NuGet.Core.Tests/NuGet.Configuration.Test/PackageNamespacesProviderTests.cs

```diff
@@ -0,0 +1,423 @@
+        }
+
+        [Fact]
+        public void AddOrUpdatePackageSourceNamespace_WithUpdatedNamespace()
+        {
+            // Arrange
+            using var mockBaseDirectory = TestDirectory.Create();
+            var configPath1 = Path.Combine(mockBaseDirectory, "NuGet.Config");
+            SettingsTestUtils.CreateConfigurationFile(configPath1, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <packageSource key=""nuget.org"">
+            <namespace id=""stuff"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+            var configPath2 = Path.Combine(mockBaseDirectory, "NuGet.Config.2");
+            SettingsTestUtils.CreateConfigurationFile(configPath2, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <packageSource key=""contoso"">
+            <namespace id=""stuff2"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+            var settings = Settings.LoadSettingsGivenConfigPaths(new string[] { configPath1, configPath2 });
+
+            // Act & Assert
+            var namespaceProvider = new PackageNamespacesProvider(settings);
+            var packageSourceNamespaces = namespaceProvider.GetPackageSourceNamespaces();
+            var namespaceToUpdate = packageSourceNamespaces.Last();
+            namespaceToUpdate.Namespaces.Add(new NamespaceItem("added"));
+            namespaceProvider.AddOrUpdatePackageSourceNamespace(namespaceToUpdate);
+            var result = @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <packageSource key=""nuget.org"">
+            <namespace id=""stuff"" />
+            <namespace id=""added"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>";
+
+            result.Replace("\r\n", "\n")
+                .Should().BeEquivalentTo(
+                File.ReadAllText(configPath1).Replace("\r\n", "\n"));
+        }
+
+        [Fact]
+        public void AddOrUpdatePackageSourceNamespace_WithANewNamespace_AddInFurthestConfig()
+        {
+            // Arrange
+            using var mockBaseDirectory = TestDirectory.Create();
+            var configPath1 = Path.Combine(mockBaseDirectory, "NuGet.Config");
+            SettingsTestUtils.CreateConfigurationFile(configPath1, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <packageSource key=""nuget.org"">
+            <namespace id=""stuff"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+            var configPath2 = Path.Combine(mockBaseDirectory, "NuGet.Config.2");
+            SettingsTestUtils.CreateConfigurationFile(configPath2, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <packageSource key=""contoso"">
+            <namespace id=""stuff2"" />
```

Comment:
Is there any unit test does `remove namespace by id`?

Reply:
From an API perspective, it's not possible to remove a namespace by id. 

The namespace item is owned by the package source namespace item, so that item knows what's available.

That pattern is established in https://github.com/NuGet/NuGet.Client/blob/dev-nkolev92-namespacesConfiguration/src/NuGet.Core/NuGet.Configuration/Settings/SettingItem.cs